From: =?utf-8?q?Lukas_M=C3=A4rdian?= <slyon@ubuntu.com>
Date: Wed, 7 Dec 2022 11:26:19 +0100
Subject: generate:util: fix double-slash root filepath

Make sure we're using proper paths values (i.e. NOT "//etc/netplan/...")
as we're doing string comparisons on those values and the leading
double-slash might confuse those checks.

E.g. in netplan.c:netplan_state_write_yaml_file() we compare netdef->filepath
to the filepath of a new YAML file to be written. If a netdef has been parsed
but its filepath is using double slashes, it won't match the new file and the
new file will be deleted instead of updated.

Starting with an empty Netplan config in /etc/netplan:
$ netplan set --origin-hint test "bridges.br54.dhcp4=false"
=> will create test.yaml, containing the br54 definition
$ netplan set --origin-hint test "bridges.br54.dhcp4=true"
=> updating the same br54 netdef (re-using 'netplan set') will delete test.yaml
---
 src/generate.c | 6 ++++--
 src/util.c     | 4 +++-
 2 files changed, 7 insertions(+), 3 deletions(-)

diff --git a/src/generate.c b/src/generate.c
index 906799d..52f0ac2 100644
--- a/src/generate.c
+++ b/src/generate.c
@@ -319,8 +319,10 @@ int main(int argc, char** argv)
             start_unit_jit("systemd-networkd-wait-online.service");
             start_unit_jit("systemd-networkd.service");
         }
-        g_autofree char* glob_run = g_strjoin(NULL, rootdir ?: "", G_DIR_SEPARATOR_S,
-                                              "run/systemd/system/netplan-*.service", NULL);
+        g_autofree char* glob_run = g_build_path(G_DIR_SEPARATOR_S,
+                                                 rootdir ?: G_DIR_SEPARATOR_S,
+                                                 "run/systemd/system/netplan-*.service",
+                                                 NULL);
         if (!glob(glob_run, 0, NULL, &gl)) {
             for (size_t i = 0; i < gl.gl_pathc; ++i) {
                 gchar *unit_name = g_path_get_basename(gl.gl_pathv[i]);
diff --git a/src/util.c b/src/util.c
index 6e75cf5..5955b3f 100644
--- a/src/util.c
+++ b/src/util.c
@@ -112,7 +112,9 @@ unlink_glob(const char* rootdir, const char* _glob)
 int find_yaml_glob(const char* rootdir, glob_t* out_glob)
 {
     int rc;
-    g_autofree char* rglob = g_strjoin(NULL, rootdir ?: "", G_DIR_SEPARATOR_S, "{lib,etc,run}/netplan/*.yaml", NULL);
+    g_autofree char* rglob = g_build_path(G_DIR_SEPARATOR_S,
+                                          rootdir ?: G_DIR_SEPARATOR_S,
+                                          "{lib,etc,run}/netplan/*.yaml", NULL);
     rc = glob(rglob, GLOB_BRACE, NULL, out_glob);
     if (rc != 0 && rc != GLOB_NOMATCH) {
         // LCOV_EXCL_START
